chain/testing.go

```diff
@@ -74,64 +106,108 @@ func (f *Builder) BuildOn(parent *types.Block, build func(b *BlockBuilder)) *typ
+	require.True(f.t, height > 0, "")
+	for i := 0; i < height; i++ {
+		parent = f.BuildOnTip(parent, 1, singleBuilder(build))
 	}
-	return parent.At(0)
+	return parent
 }
 
-// BuildOnTip creates and returns a new block child of `parent`.
+// BuildOnTip creates and returns a new tipset child of `parent`.
+// The tipset carries `width` > 0 blocks with the same height and parents, but different tickets.
+// Note: the blocks will all have the same miner, which is unrealistic and forbidden by consensus;
+// generalise this to random miner addresses when that is rejected by the syncer.
 // The `build` function is invoked to modify the block before it is stored.
-func (f *Builder) BuildOnTip(parent types.TipSet, build func(b *BlockBuilder)) *types.Block {
-	ticket := make([]byte, binary.Size(f.seq))
-	binary.BigEndian.PutUint64(ticket, f.seq)
-	f.seq++
-
-	// Sum weight of parents' parent weight, plus one for each parent.
-	// Note: as with the state builder, we should probably factor this out.
-	parentWeight := types.Uint64(0)
-	for i := 0; i < parent.Len(); i++ {
-		parentWeight += parent.At(i).ParentWeight + 1
-	}
+func (f *Builder) BuildOnTip(parent types.TipSet, width int, build func(b *BlockBuilder, i int)) types.TipSet {
+	require.True(f.t, width > 0)
+	var blocks []*types.Block
 
 	height := types.Uint64(0)
+	grandparentKey := types.NewTipSetKey()
 	if parent.Defined() {
+		var err error
 		height = parent.At(0).Height + 1
+		grandparentKey, err = parent.Parents()
+		require.NoError(f.t, err)
 	}
 
-	b := &types.Block{
-		Ticket:          ticket,
-		Miner:           f.minerAddress,
-		ParentWeight:    parentWeight,
-		Parents:         parent.Key(),
-		Height:          height,
-		Messages:        []*types.SignedMessage{},
-		MessageReceipts: []*types.MessageReceipt{},
-		// Omitted fields below
-		//StateRoot:       stateRoot,
-		//Proof            PoStProof
-		//Timestamp        Uint64
-	}
-	// Nonce intentionally omitted as it will go away.
+	parentWeight, err := f.stateBuilder.Weigh(parent, f.StateForKey(grandparentKey))
+	require.NoError(f.t, err)
+
+	for i := 0; i < width; i++ {
+		ticket := make([]byte, binary.Size(f.seq))
+		binary.BigEndian.PutUint64(ticket, f.seq)
+		f.seq++
+
+		b := &types.Block{
+			Ticket:          ticket,
+			Miner:           f.minerAddress,
+			ParentWeight:    types.Uint64(parentWeight),
+			Parents:         parent.Key(),
+			Height:          height,
+			Messages:        []*types.SignedMessage{},
+			MessageReceipts: []*types.MessageReceipt{},
+			// Omitted fields below
+			//StateRoot:       stateRoot,
+			//Proof            PoStProof
+			//Timestamp        Uint64
+		}
+		// Nonce intentionally omitted as it will go away.
+
+		if build != nil {
+			build(&BlockBuilder{b}, i)
+		}
 
-	if build != nil {
-		build(&BlockBuilder{b})
+		// Compute state root for this block.
+		tipSet, err := types.NewTipSet(b)
+		require.NoError(f.t, err)
+		b.StateRoot = f.ComputeState(tipSet)
```

Comment:
I'm a bit worried that we are treating the `builder` like a chain store (it implements the store interface that the syncer needs) and not keeping track of intermediate state roots.  As it stands right now the `builder.GetTipSetStateRoot` method will error in places the `GetTipSet` method will not because we track intermediate blocks and not intermediate state roots.